README.md
# JsonDecoder

[![Build Status](https://travis-ci.org/joanllenas/ts.data.json.svg?branch=master)](https://travis-ci.org/joanllenas/ts.data.json)
[![npm version](https://badge.fury.io/js/ts.data.json.svg)](https://www.npmjs.com/package/ts.data.json)

Typescript type annotations give us compile-time guarantees, but at run-time, when data flows from the server to our clients, lots of things can go wrong.

JSON decoders validate the JSON before it comes into our program. So if the data has an unexpected structure, we learn about it immediately.

> If you are new to JSON decoding, you may want to read the introductory article [Decoding JSON with Typescript](https://dev.to/joanllenas/decoding-json-with-typescript-1jjc) about why and how to use this library.

[![](./.github/all-your-json-are-belong-to-us.jpg)](https://en.wikipedia.org/wiki/All_your_base_are_belong_to_us)

## Install

```
npm install ts.data.json --save
```

## Example

```ts
type User = {
  firstname: string;
  lastname: string;
};

const userDecoder = JsonDecoder.object<User>(
  {
    firstname: JsonDecoder.string,
    lastname: JsonDecoder.string
  },
  'User'
);

const jsonObjectOk = {
  firstname: 'Damien',
  lastname: 'Jurado'
};

userDecoder
  .decodePromise(jsonObjectOk)
  .then(user => {
    console.log(`User ${user.firstname} ${user.lastname} decoded successfully`);
  })
  .catch(error => {
    console.log(error);
  });

// Output: User Damien Jurado decoded successfully

const jsonObjectKo = {
  firstname: 'Erik',
  lastname: null
};

userDecoder
  .decodePromise(jsonObjectKo)
  .then(user => {
    console.log('User decoded successfully');
  })
  .catch(error => {
    console.error(error);
  });

// Output: <User> decoder failed at key "lastname" with error: null is not a valid string
```

## API

### JsonDecoder.string

> `string: Decoder<string>`

Creates a `string` decoder.

```ts
JsonDecoder.string.decode('hi'); // Ok<string>({value: 'hi'})
JsonDecoder.string.decode(5); // Err({error: '5 is not a valid string'})
```

### JsonDecoder.number

> `number: Decoder<number>`

Creates a `number` decoder.

```ts
JsonDecoder.number.decode(99); // Ok<number>({value: 99})
JsonDecoder.number.decode('hola'); // Err({error: 'hola is not a valid number'})
```

### JsonDecoder.boolean

> `boolean: Decoder<boolean>`

Creates a `boolean` decoder.

```ts
JsonDecoder.boolean.decode(true); // Ok<boolean>({value: true})
JsonDecoder.boolean.decode(null); // Err({error: 'null is not a valid boolean'})
```

### JsonDecoder.object

> `object<a>(decoders: DecoderObject<a>, decoderName: string, keyMap?: DecoderObjectKeyMap<a>): Decoder<a>`

Creates an `object` decoder.

#### @param `decoders: DecoderObject<a>`

Key/value pair that has to comply with the `<a>` type.

> Turns all optional keys to required, so you have to specify decoders even for the optional (i.e. with `{name?: string}`) keys.

#### @param `decoderName: string`

Type of the object we are decoding. i.e. `User`. It is used to generate meaningful decoding error messages.

#### @param `keyMap?: DecoderObjectKeyMap<a>`

Optional key/value pair to map JSON-land keys with Model-land keys.
Useful when the JSON keys don't match with the decoded type keys.

#### Basic example

```ts
type User = {
  firstname: string;
  lastname: string;
};
const userDecoder = JsonDecoder.object<User>(
  {
    firstname: JsonDecoder.string,
    lastname: JsonDecoder.string
  },
  'User'
);

const jsonOk = {
  firstname: 'Damien',
  lastname: 'Jurado'
};
userDecoder.decode(jsonOk);
// Output: Ok<User>({value: {firstname: 'Damien', lastname: 'Jurado'}})

const jsonKo = {
  firstname: null,
  lastname: 'Satie'
};
userDecoder.decode(jsonKo);
// Output: Err({error: '<User> decoder failed at key "firstname" with error: null is not a valid string'})
```

#### keyMap example

```ts
const userDecoder = JsonDecoder.object<User>(
  {
    firstname: JsonDecoder.string,
    lastname: JsonDecoder.string
  },
  'User',
  {
    firstname: 'fName',
    lastname: 'lName'
  }
);

const jsonOk = {
  fName: 'Nick',
  lName: 'Drake'
};
userDecoder.decode(json);
// Output: Ok({value: {firstname: 'Nick', lastname: 'Drake'}})

const jsonKo = {
  fName: 'Nick'
};
userDecoder.decode(json);
// Output: Err({error: '<User> decoder failed at key "lastname" (mapped from the JSON key "lName") with error: undefined is not a valid string'})
```

### JsonDecoder.objectStrict

> `objectStrict<a>(decoders: DecoderObject<a>, decoderName: string): Decoder<a>`

Creates an `object` decoder that performs strict key checks. It only accepts json objects with exactly the same keys as the decoder keys.

#### @param `decoders: DecoderObject<a>`

Key/value pair that has to comply with the `<a>` type.

#### @param `decoderName: string`

Type of the object we are decoding. i.e. `User`. It is used to generate meaningful decoding error messages.

#### Basic example

```ts
type User = {
  firstname: string;
  lastname: string;
};
const userDecoder = JsonDecoder.objectStrict<User>(
  {
    firstname: JsonDecoder.string,
    lastname: JsonDecoder.string
  },
  'User'
);

const jsonOk = {
  firstname: 'Damien',
  lastname: 'Jurado'
};
userDecoder.decode(jsonOk);
// Output: Ok<User>({value: {firstname: 'Damien', lastname: 'Jurado'}})

const jsonKo = {
  firstname: 'Damien',
  lastname: 'Jurado',
  email: 'damien@damienjurado.com'
};
userDecoder.decode(jsonKo);
// Output: Err({error: 'Unknown key "email" found while processing strict <User> decoder'})
```

### JsonDecoder.array

> `array<a>(decoder: Decoder<a>, decoderName: string): Decoder<Array<a>>`

Creates an `array` decoder.

#### @param `decoder: Decoder<a>`

The decoder used to decode every `Array<a>` item.

#### @param `decoderName: string`

Type of the object we are decoding. i.e. `User[]`. It is used to generate meaningful decoding error messages.

```ts
JsonDecoder.array<number>(JsonDecoder.number, 'number[]').decode([1, 2, 3]);
// Output: Ok<number[]>({value: [1, 2, 3]})

JsonDecoder.array<number>(JsonDecoder.number, 'number[]').decode([1, '2', 3]);
// Output: Err({error: '<number[]> decoder failed at index 1 with error: "2" is not a valid number'})
```

### JsonDecoder.dictionary

> `dictionary<a>(decoder: Decoder<a>, decoderName: string): Decoder<{ [name: string]: a }>`

Creates a `dictionary` decoder.

#### @param `decoder: Decoder<a>`

The decoder used to decode every value of the key/value pairs.

#### @param `decoderName: string`

Type of the object we are decoding. i.e. `User`. It is used to generate meaningful decoding error messages.

```ts
JsonDecoder.dictionary(JsonDecoder.number, 'Dict<number>').decode({
  a: 1,
  b: 2
});
// Output: Ok<Dict<number>>({value: {a: 1, b: 2}})

JsonDecoder.dictionary(JsonDecoder.number, 'Dict<number>').decode({
  a: 1,
  b: 2,
  c: null
});
// Output: Err({error: '<Dict<number>> dictionary decoder failed at key "c" with error: null is not a valid number'})
```

### JsonDecoder.oneOf

> `oneOf<a>(decoders: Array<Decoder<a>>, decoderName: string): Decoder<a>`

The `oneOf` decoder tries to decode the provided JSON with any of the provided decoders. It returns `Ok` with the first successful decoded value or `Err` if all decoders fail.

#### @param `decoders: Array<Decoder<a>>`

The Array of decoders the JSON can be decoded with.

#### @param `decoderName: string`

Type of the object we are decoding. i.e. `number | string`. It is used to generate meaningful decoding error messages.

```ts
JsonDecoder.oneOf<string | number>(
  [JsonDecoder.string, JsonDecoder.number],
  'string | number'
).decode(1);
// Output: Ok<string | number>({value: 1})

JsonDecoder.oneOf<string | number>(
  [JsonDecoder.string, JsonDecoder.number],
  'string | number'
).decode(true);
// Output: Err({error: "<string | number> decoder failed because true can't be decoded with any of the provided oneOf decoders"})
```

### JsonDecoder.allOf

> `allOf<T extends Array<Decoder<unknown>>, R = AllOfDecoderReturn<T>>(decoders: T): Decoder<R>`

The `allOf` decoder tries to decode the provided JSON with all of the provided decoders, in order. The output of one decoder is passed as input to the next decoder. It returns `Ok` with the last successful decoded value or `Err` if any decoder fails.

The `allOf` decoder allows you to combine multiple decoders. It is probably most useful when combined with custom decoders you may make for your application.

#### @param `decoders: T extends Array<Decoder<unknown>>`

An array of decoders the JSON should be decoded with.

Simple examples:

```ts
JsonDecoder.allOf(
 JsonDecoder.string,
 JsonDecoder.failover(10, JsonDecoder.number)
).decode('hola'),
// Output: Ok({value: 10})

JsonDecoder.allOf(
 JsonDecoder.string,
 JsonDecoder.failover(10, JsonDecoder.number)
).decode(5),
// Output: Err({error: "5 is not a valid string})
```

Example using a custom `hasLength()` decoder (an example [here](https://stackblitz.com/edit/typescript-97ergz)):

```ts
JsonDecoder.allOf(
  JsonDecoder.array(JsonDecoder.number, 'latLang'),
  hasLength<[number, number]>(2)
).decode([-123.34324, 23.454365]);
// Output: Ok({value: [-123.34324, 23.454365]})

JsonDecoder.allOf(
  JsonDecoder.array(JsonDecoder.number, 'latLang'),
  hasLength<[number, number]>(2)
).decode([1, 2, 3]);
// Output: Err({error: "hasLength() decoder failed because the provided array is of length 3."})
```

### JsonDecoder.lazy

> `lazy<a>(mkDecoder: () => Decoder<a>): Decoder<a>`

Decoder for recursive data structures.

#### @param `mkDecoder: () => Decoder<a>`

A function that returns a decoder.

```ts
type Node<a> = {
  value: a;
  children?: Node<a>[];
};
const treeDecoder: JsonDecoder.Decoder<Node<string>> = JsonDecoder.object<
  Node<string>
>(
  {
    value: JsonDecoder.string,
    children: JsonDecoder.oneOf<Node<string>[]>(
      [
        JsonDecoder.lazy(() => JsonDecoder.array(treeDecoder, 'Node<a>[]')),
        JsonDecoder.isUndefined([])
      ],
      'Node<string>[] | isUndefined'
    )
  },
  'Node<string>'
);
treeDecoder.decode({
  value: 'root',
  children: [
    { value: '1' },
    { value: '2', children: [{ value: '2.1' }, { value: '2.2' }] }
  ]
});
// Output: Ok<Node<string>>({value: {value: 'root', children: [....]}})

treeDecoder.decode({
  value: 'root',
  children: null
});
// Output: Err({error: "<Node<string>> decoder failed at key 'children' with error: <Node<string>[] | isUndefined> decoder failed because null can't be decoded with any of the provided oneOf decoders"})
```

### JsonDecoder.optional

> `optional<a>(decoder: Decoder<a>): Decoder<a | undefined>`

The `optional` decoder tries to decode the provided JSON with the provided decoder if the json value is not `undefined` or `null`. This decoder is to allow for an optional value in the TypeScript definition while retaining the ability to give a detailed error message if the wrapped decoder fails.

#### @param `decoder: Decoder<a>`

Decoder the JSON will be decoded with if the value is not `null` or `undefined`.

```ts
type User = {
  firstname: string;
  lastname: string;
  email?: string;
};
const userDecoder = JsonDecoder.object<User>(
  {
    firstname: JsonDecoder.string,
    lastname: JsonDecoder.string,
    email: JsonDecoder.optional(JsonDecoder.string)
  },
  'User'
);

const jsonOk = {
  firstname: 'Damien',
  lastname: 'Jurado'
};

const jsonFullUser = {
  firstname: 'Damien',
  lastname: 'Jurado',
  email: 'user@example.com'
};

const jsonKo = {
  firstname: null,
  lastname: 'Satie'
};

JsonDecoder.optional(userDecoder).decode(null);
// Output: Ok<User | undefined>({value: undefined})

JsonDecoder.optional(userDecoder).decode(undefined);
// Output: Ok<User | undefined>({value: undefined})

JsonDecoder.optional(userDecoder).decode(jsonOk);
// Output: Ok<User | undefined>({value: {firstname: 'Damien', lastname: 'Jurado', email: undefined}})

JsonDecoder.optional(userDecoder).decode(jsonFullUser);
// Output: Ok<User | undefined>({value: {firstname: 'Damien', lastname: 'Jurado', email: 'user@example.com'}})

JsonDecoder.optional(userDecoder).decode(jsonKo);
// Output: Err({error: '<User> decoder failed at key "firstname" with error: null is not a valid string'})
```

### JsonDecoder.failover

> `failover<a>(defaultValue: a, decoder: Decoder<a>): Decoder<a>`

Creates a decoder that returns a default value on failure.

#### @param `defaultValue: a`

The `Ok` default value when the decoder fails.

#### @param `decoder: Decoder<a>`

Decoder the JSON will be decoded with.

```ts
JsonDecoder.failover('default value', JsonDecoder.string).decode(
  'This is fine'
);
// Ok<string>({value: 'This is fine'})

JsonDecoder.failover('default value', JsonDecoder.string).decode(null);
// Ok<string>({value: 'default value'})
```

### JsonDecoder.succeed

> `succeed: Decoder<any>`

Creates a decoder that always succeeds.

```ts
JsonDecoder.succeed.decode(null); // Ok<any>({value: null})
```

### JsonDecoder.fail

> `fail<a>(error: string): Decoder<a>`

Creates a decoder that always fails.

#### @param `error: string`

Error message that will be returned with the `Err` instance.

```ts
JsonDecoder.fail('Something wrong happened').decode('This is fine');
// Err({error: 'Something wrong happened'})
```

### JsonDecoder.isNull

> `isNull<a>(defaultValue: a): Decoder<a>`

Succeeds when JSON is strictly (===) null and returns a defaultValue.

#### @param `defaultValue: a`

Returned default value when JSON is null.

```ts
JsonDecoder.isNull('default value').decode(null);
// Ok({value: 'default value'})

JsonDecoder.isNull('default value').decode(999);
// Err({error: '999 is not null'})
```

### JsonDecoder.isUndefined

> `isUndefined<a>(defaultValue: a): Decoder<a>`

Succeeds when JSON is strictly (===) undefined and returns a defaultValue.

#### @param `defaultValue: a`

Returned default value when JSON is undefined.

```ts
JsonDecoder.isUndefined('default value').decode(undefined);
// Ok({value: 'default value'})

JsonDecoder.isUndefined('default value').decode(999);
// Err({error: '999 is not undefined'})
```

### JsonDecoder.isExactly

> `isExactly<a>(value: a): Decoder<a>`

Succeeds when JSON is strictly (===) `value: a` and returns `value: a`.

#### @param `value: a`

Value returned when the JSON is strictly equal to it.

```ts
JsonDecoder.isExactly(true).decode(true);
// Ok({value: true})

JsonDecoder.isExactly(999).decode(true);
// Err({error: 'true is not 999'})
```

### JsonDecoder.constant

> `constant<a>(value: a): Decoder<a>`

A Decoder that always succeeds, returning `value`.

#### @param `value: a`

Value always returned.

```ts
JsonDecoder.constant(true).decode(false);
// Ok({value: true})
```

## Related libraries

- https://github.com/gcanti/io-ts
- https://github.com/kofno/jsonous
- https://github.com/jquense/yup
- https://gitlab.com/john.carroll.p/ts-decoders


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.2.1"></a>
## [0.2.1](https://github.com/joanllenas/ts.data.json/compare/v0.2.0...v0.2.1) (2019-07-23)



<a name="0.2.0"></a>
# [0.2.0](https://github.com/joanllenas/ts.data.json/compare/v0.1.0...v0.2.0) (2019-05-18)


### Features

* **Result:** Result is a Functor now ([356c3b0](https://github.com/joanllenas/ts.data.json/commit/356c3b0))



<a name="0.1.0"></a>
# [0.1.0](https://github.com/joanllenas/ts.data.json/compare/v0.0.8...v0.1.0) (2019-05-06)


### Features

* **allOf:** add allOf decoder ([aa9f31b](https://github.com/joanllenas/ts.data.json/commit/aa9f31b))



<a name="0.0.8"></a>
## [0.0.8](https://github.com/joanllenas/ts.data.json/compare/v0.0.7...v0.0.8) (2019-01-18)



<a name="0.0.7"></a>
## [0.0.7](https://github.com/joanllenas/ts.data.json/compare/v0.0.6...v0.0.7) (2018-12-23)


### Bug Fixes

* **readme:** Improved readme docs. Improved error reporting ([503e2e5](https://github.com/joanllenas/ts.data.json/commit/503e2e5))



<a name="0.0.6"></a>
## [0.0.6](https://github.com/joanllenas/ts.data.json/compare/v0.0.5...v0.0.6) (2018-12-22)


### Bug Fixes

* **JsonDecoder.any:** Renamed to succeedfor consistency ([414b87a](https://github.com/joanllenas/ts.data.json/commit/414b87a))



<a name="0.0.5"></a>
## [0.0.5](https://github.com/joanllenas/ts.data.json/compare/v0.0.4...v0.0.5) (2018-12-22)


### Bug Fixes

* **JsonDecoder.array:** Improved error reporting ([adf6adc](https://github.com/joanllenas/ts.data.json/commit/adf6adc))



<a name="0.0.4"></a>
## [0.0.4](https://github.com/joanllenas/ts.data.json/compare/v0.0.3...v0.0.4) (2018-12-21)



<a name="0.0.3"></a>
## 0.0.3 (2018-12-21)


### Bug Fixes

* **project:** Project rename ([5fb23c1](https://github.com/joanllenas/ts.data.json/commit/5fb23c1))



# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.


src/result.spec.ts
import { JsonDecoder, $JsonDecoderErrors } from './json-decoder';

import * as chai from 'chai';
import { Ok, ok, err } from './result';

const expect = chai.expect;

describe('Result', () => {
  const toUpper = (str: string) => str.toUpperCase();
  const exclamation = (str: string) => `${str}!`;

  describe('Ok', () => {
    it('should wrap the value', () => {
      expect(new Ok('hola').value).to.equal('hola');
    });
    it('should chain transformations while mapping', () => {
      expect(
        ok('hola')
          .map(toUpper)
          .map(exclamation)
      ).to.deep.equal(ok('HOLA!'));
    });
  });

  describe('Err', () => {
    it('should wrap the error', () => {
      expect(err('Wrong!')).to.deep.equal(err('Wrong!'));
    });
    it('should keep the first error while mapping', () => {
      expect(
        err<string>('Wrong')
          .map(toUpper)
          .map(exclamation)
      ).to.deep.equal(err('Wrong'));
    });
  });

  it('should allow chaining more than one transformation', () => {
    const onlyEven = (num: number) =>
      num % 2 === 0 ? ok(num) : err<number>('not even');
    const example = (n: number) =>
      onlyEven(n)
        .map(n2 => n2 * 2)
        .map(n2 => Number(n2).toString(16))
        .map(hex => `#${hex}`);
    expect(example(5)).to.deep.equal(err('not even'));
    expect(example(6)).to.deep.equal(ok('#c'));
  });
});


src/json-decoder.ts
import { Result, Ok, ok, err } from './result';

export namespace JsonDecoder {
  export class Decoder<a> {
    constructor(private decodeFn: (json: any) => Result<a>) {}

    /**
     * Decodes a JSON object of type <a> and returns a Result<a>
     * @param json The JSON object
     */
    decode(json: any): Result<a> {
      return this.decodeFn(json);
    }

    /**
     * Decodes a JSON object of type <a> and calls onOk() on success or onErr() on failure, both return <b>
     * @param json The JSON object to decode
     * @param onOk function called when the decoder succeeds
     * @param onErr function called when the decoder fails
     */
    onDecode<b>(
      json: any,
      onOk: (result: a) => b,
      onErr: (error: string) => b
    ): b {
      const result = this.decode(json);
      if (result instanceof Ok) {
        return onOk(result.value);
      } else {
        return onErr(result.error);
      }
    }

    /**
     * Decodes a JSON object of type <a> and returns a Promise<a>
     * @param json The JSON object to decode
     */
    decodePromise<b>(json: any): Promise<a> {
      return new Promise<a>((resolve, reject) => {
        const result = this.decode(json);
        if (result instanceof Ok) {
          return resolve(result.value);
        } else {
          return reject(result.error);
        }
      });
    }

    /**
     * Chains decoder result transformations
     * @param fn The transformation function
     */
    map<b>(fn: (value: a) => b): Decoder<b> {
      return new Decoder<b>((json: any) => {
        const result = this.decodeFn(json);
        if (result instanceof Ok) {
          return ok(fn(result.value));
        } else {
          return err(result.error);
        }
      });
    }

    /**
     * Chains decoders
     * @param fn Function that returns a new decoder
     */
    then<b>(fn: (value: a) => Decoder<b>): Decoder<b> {
      return new Decoder<b>((json: any) => {
        const result = this.decodeFn(json);
        if (result instanceof Ok) {
          return fn(result.value).decode(json);
        } else {
          return err(result.error);
        }
      });
    }
  }

  /**
   * Decoder for recursive data structures.
   *
   * @param mkDecoder A function that returns a decoder
   */
  export function lazy<a>(mkDecoder: () => Decoder<a>): Decoder<a> {
    return new Decoder((json: any) => mkDecoder().decode(json));
  }

  /**
   * Decoder for `string`.
   */
  export const string: Decoder<string> = new Decoder<string>((json: any) => {
    if (typeof json === 'string') {
      return ok<string>(json);
    } else {
      return err<string>($JsonDecoderErrors.primitiveError(json, 'string'));
    }
  });

  /**
   * Decoder for `number`.
   */
  export const number: Decoder<number> = new Decoder<number>((json: any) => {
    if (typeof json === 'number') {
      return ok<number>(json);
    } else {
      return err<number>($JsonDecoderErrors.primitiveError(json, 'number'));
    }
  });

  /**
   * Decoder for `boolean`.
   */
  export const boolean: Decoder<boolean> = new Decoder<boolean>((json: any) => {
    if (typeof json === 'boolean') {
      return ok<boolean>(json);
    } else {
      return err<boolean>($JsonDecoderErrors.primitiveError(json, 'boolean'));
    }
  });

  export type DecoderObject<a> = { [p in keyof Required<a>]: Decoder<a[p]> };
  export type DecoderObjectKeyMap<a> = { [p in keyof a]?: string };

  /**
   * Decoder for objects.
   *
   * @param decoders Key/value pairs of decoders for each object field.
   * @param decoderName How to display the name of the object being decoded in errors.
   * @param keyMap Optional map between json field names and user land field names.
   *               Useful when the client model does not match with what the server sends.
   */
  export function object<a>(
    decoders: DecoderObject<a>,
    decoderName: string,
    keyMap?: DecoderObjectKeyMap<a>
  ): Decoder<a> {
    return new Decoder<a>((json: any) => {
      if (json !== null && typeof json === 'object') {
        const result: any = {};
        for (const key in decoders) {
          if (decoders.hasOwnProperty(key)) {
            if (keyMap && key in keyMap) {
              const jsonKey = keyMap[key] as string;
              const r = decoders[key].decode(json[jsonKey]);
              if (r instanceof Ok) {
                result[key] = r.value;
              } else {
                return err<a>(
                  $JsonDecoderErrors.objectJsonKeyError(
                    decoderName,
                    key,
                    jsonKey,
                    r.error
                  )
                );
              }
            } else {
              const r = decoders[key].decode(json[key]);
              if (r instanceof Ok) {
                result[key] = r.value;
              } else {
                return err<a>(
                  $JsonDecoderErrors.objectError(decoderName, key, r.error)
                );
              }
            }
          }
        }
        return ok<a>(result);
      } else {
        return err<a>($JsonDecoderErrors.primitiveError(json, decoderName));
      }
    });
  }

  /**
   * Decoder for objects that performs strict key checks.
   * The decoder will fail if there are any extra keys in the provided object.
   *
   * @param decoders Key/value pairs of decoders for each object field.
   * @param decoderName How to display the name of the object being decoded in errors.
   */
  export function objectStrict<a>(
    decoders: DecoderObject<a>,
    decoderName: string
  ): Decoder<a> {
    return new Decoder<a>((json: any) => {
      if (json !== null && typeof json === 'object') {
        for (const key in json) {
          if (!decoders.hasOwnProperty(key)) {
            return err<a>(
              $JsonDecoderErrors.objectStrictUnknownKeyError(decoderName, key)
            );
          }
        }
        const result: any = {};
        for (const key in decoders) {
          if (decoders.hasOwnProperty(key)) {
            const r = decoders[key].decode(json[key]);
            if (r instanceof Ok) {
              result[key] = r.value;
            } else {
              return err<a>(
                $JsonDecoderErrors.objectError(decoderName, key, r.error)
              );
            }
          }
        }
        return ok<a>(result);
      } else {
        return err<a>($JsonDecoderErrors.primitiveError(json, decoderName));
      }
    });
  }

  /**
   * Always succeeding decoder
   */
  export const succeed: Decoder<any> = new Decoder<any>((json: any) => {
    return ok<any>(json);
  });

  /**
   * Always failing decoder
   */
  export function fail<a>(error: string): Decoder<a> {
    return new Decoder<a>((json: any) => {
      return err<any>(error);
    });
  }

  /**
   * Tries to decode with `decoder` and returns `defaultValue` on failure.
   * (It was called maybe() before)
   *
   * @param defaultValue The default value returned in case of decoding failure.
   * @param decoder The actual decoder to use.
   */
  export function failover<a>(
    defaultValue: a,
    decoder: Decoder<a>
  ): Decoder<a> {
    return new Decoder<a>((json: any) => {
      const result = decoder.decode(json);
      if (result instanceof Ok) {
        return result;
      } else {
        return ok<a>(defaultValue);
      }
    });
  }

  /**
   * Tries to decode with `decoder` and returns `error` on failure, but allows
   * for `undefined` or `null` values to be present at the top level and returns
   * an `undefined` if the value was `undefined` or `null`.
   *
   * @param decoder The actual decoder to use.
   */
  export function optional<a>(decoder: Decoder<a>): Decoder<a | undefined> {
    return new Decoder<a | undefined>((json: any) => {
      if (json === undefined) {
        return ok<undefined>(undefined);
      } else if (json === null) {
        return ok<undefined>(undefined);
      } else {
        return decoder.decode(json);
      }
    });
  }

  /**
   * Tries to decode the provided json value with any of the provided `decoders`.
   * If all provided `decoders` fail, this decoder fails.
   * Otherwise, it returns the first successful decoder.
   *
   * @param decoders An array of decoders to try.
   */
  export function oneOf<a>(
    decoders: Array<Decoder<a>>,
    decoderName: string
  ): Decoder<a> {
    return new Decoder<a>((json: any) => {
      for (let i = 0; i < decoders.length; i++) {
        const result = decoders[i].decode(json);
        if (result instanceof Ok) {
          return result;
        }
      }
      return err<a>($JsonDecoderErrors.oneOfError(decoderName, json));
    });
  }

  type SubtractOne<T extends number> = [
    -1,
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
    19,
    20,
    21,
    22,
    23,
    24,
    25,
    26,
    27,
    28,
    29,
    30
  ][T];

  /**
   * Plucks the last type in a tuple of length 30 or less.
   * Else returns the first type in a tuple.
   */
  export type AllOfDecoderReturn<T extends unknown[]> = T[SubtractOne<
    T['length']
  >] extends JsonDecoder.Decoder<infer R>
    ? R
    : T[0];

  /**
   * Tries to decode the provided json value with all of the provided `decoders`.
   * The order of the provided decoders matters: the output of one decoder is passed
   * as input to the next decoder. If any of the provided `decoders` fail, this
   * decoder fails. Otherwise, it returns the output of the last decoder.
   *
   * @param decoders a spread of decoders to use.
   */
  export function allOf<
    T extends Array<Decoder<unknown>>,
    R = AllOfDecoderReturn<T>
  >(...decoders: T): Decoder<R> {
    return new Decoder<R>((json: any) =>
      decoders.reduce(
        (prev, curr) =>
          (prev instanceof Ok ? curr.decode(prev.value) : prev) as Result<R>,
        ok<R>(json)
      )
    );
  }

  /**
   * Decoder for key/value pairs.
   *
   * @param decoder An object decoder for the values. All values must have the same shape or use oneOf otherwise.
   */
  export const dictionary = <a>(
    decoder: Decoder<a>,
    decoderName: string
  ): Decoder<{ [name: string]: a }> => {
    return new Decoder<{ [name: string]: a }>(json => {
      if (json !== null && typeof json === 'object') {
        const obj: { [name: string]: a } = {};
        for (const key in json) {
          if (json.hasOwnProperty(key)) {
            const result = decoder.decode(json[key]);
            if (result instanceof Ok) {
              obj[key] = result.value;
            } else {
              return err<{ [name: string]: a }>(
                $JsonDecoderErrors.dictionaryError(
                  decoderName,
                  key,
                  result.error
                )
              );
            }
          }
        }
        return ok<{ [name: string]: a }>(obj);
      } else {
        return err<{ [name: string]: a }>(
          $JsonDecoderErrors.primitiveError(json, 'dictionary')
        );
      }
    });
  };

  /**
   * Decoder for Array<T>.
   *
   * @param decoder The decoder for the array element.
   */
  export const array = <a>(
    decoder: Decoder<a>,
    decoderName: string
  ): Decoder<Array<a>> => {
    return new Decoder<Array<a>>(json => {
      if (json instanceof Array) {
        const arr: Array<a> = [];
        for (let i = 0; i < json.length; i++) {
          const result = decoder.decode(json[i]);
          if (result instanceof Ok) {
            arr.push(result.value);
          } else {
            return err<Array<a>>(
              $JsonDecoderErrors.arrayError(decoderName, i, result.error)
            );
          }
        }
        return ok<Array<a>>(arr);
      } else {
        return err<Array<a>>($JsonDecoderErrors.primitiveError(json, 'array'));
      }
    });
  };

  /**
   * Decoder that only succeeds when json is strictly (===) `null`.
   * When succeeds it returns `defaultValue`.
   *
   * @param defaultValue The value returned when json is `null`.
   */
  export function isNull<a>(defaultValue: a): Decoder<a> {
    return new Decoder((json: any) => {
      if (json === null) {
        return ok<a>(defaultValue);
      } else {
        return err<a>($JsonDecoderErrors.nullError(json));
      }
    });
  }

  /**
   * Decoder that only succeeds when json is strictly (===) `undefined`.
   * When succeeds it returns `defaultValue`.
   *
   * @param defaultValue The value returned when json is `undefined`.
   */
  export function isUndefined<a>(defaultValue: a): Decoder<a> {
    return new Decoder((json: any) => {
      if (json === undefined) {
        return ok<a>(defaultValue);
      } else {
        return err<a>($JsonDecoderErrors.undefinedError(json));
      }
    });
  }

  /**
   * Decoder that always succeeds returning `value`.
   *
   * @param value The value returned.
   */
  export const constant = <a>(value: a): Decoder<a> => {
    return new Decoder<a>((json: any) => ok(value));
  };

  /**
   * Decoder that only succeeds when json is strictly (===) `value`.
   * When succeeds it returns `value`.
   *
   * @param value The value returned on success.
   */
  export function isExactly<a>(value: a): Decoder<a> {
    return new Decoder((json: any) => {
      if (json === value) {
        return ok<a>(value);
      } else {
        return err<a>($JsonDecoderErrors.exactlyError(json, value));
      }
    });
  }
}

export namespace $JsonDecoderErrors {
  export const primitiveError = (value: any, tag: string): string =>
    `${JSON.stringify(value)} is not a valid ${tag}`;

  export const exactlyError = (json: any, value: any): string =>
    `${JSON.stringify(json)} is not exactly ${JSON.stringify(value)}`;

  export const undefinedError = (json: any): string =>
    `${JSON.stringify(json)} is not undefined`;

  export const nullError = (json: any): string =>
    `${JSON.stringify(json)} is not null`;

  export const dictionaryError = (
    decoderName: string,
    key: string,
    error: string
  ): string =>
    `<${decoderName}> dictionary decoder failed at key "${key}" with error: ${error}`;

  export const oneOfError = (decoderName: string, json: any): string =>
    `<${decoderName}> decoder failed because ${JSON.stringify(
      json
    )} can't be decoded with any of the provided oneOf decoders`;

  export const objectError = (
    decoderName: string,
    key: string,
    error: string
  ): string =>
    `<${decoderName}> decoder failed at key "${key}" with error: ${error}`;

  export const arrayError = (
    decoderName: string,
    index: number,
    error: string
  ): string =>
    `<${decoderName}> decoder failed at index "${index}" with error: ${error}`;

  export const objectJsonKeyError = (
    decoderName: string,
    key: string,
    jsonKey: string,
    error: string
  ): string =>
    `<${decoderName}> decoder failed at key "${key}" (mapped from the JSON key "${jsonKey}") with error: ${error}`;

  export const objectStrictUnknownKeyError = (
    decoderName: string,
    key: string
  ): string =>
    `Unknown key "${key}" found while processing strict <${decoderName}> decoder`;
}
